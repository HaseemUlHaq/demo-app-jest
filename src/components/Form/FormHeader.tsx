import React from 'react';
import { FormProps } from '.';

export interface HeaderProps {
  form: FormProps;
}

/** default props for Header */
export const defaultHeaderProps = {
  form: {
    name: '',
    age: 0,
  },
};

const FormHeader: React.FC<HeaderProps> = (props: HeaderProps) => {
  const { form } = props;
  return (
    <h1>
      Hello {form.name} {form.age}
    </h1>
  );
};
FormHeader.defaultProps = defaultHeaderProps;
export default FormHeader;
