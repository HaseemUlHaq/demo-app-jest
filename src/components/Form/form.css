.card {
  min-height: 400px;
  width: 35%;
  border: 2px solid #eeeeee;
  border-radius: 4px;
  margin: 80px auto;
  padding: 20px;
  text-align: left;
  box-sizing: border-box;
  box-shadow: 0 0 11px rgba(33, 33, 33, 0.2);
}
.btn-wrapper {
  margin-top: 40px;
}
.first-lt-color:before {
  color: blueviolet;
  content: 'Hello ';
}
input {
  font-size: 14px;
  padding: 0 10px;
  line-height: 2.5;
  font-weight: 500;
  border-color: blueviolet;
  border-radius: 0;
  border: 2px solid #eeeeee;
  transition: all 0.4s ease;
  width: 94%;
}
input:focus {
  outline: none;
  border-color: blueviolet;
  border-radius: 0;
  box-shadow: 0 0 11px rgba(33, 33, 33, 0.2);
}
.submit-btn {
  padding: 14px 20px;
  width: 100%;
  font-size: 17px;
  font-weight: 600;
  border-radius: 0;
  border: none;
  background-color: blueviolet;
  color: white;
  cursor: pointer;
  transition: all 0.3s ease;
}
.submit-btn:hover {
  background-color: rgb(116, 0, 224);
  box-shadow: 0 0 11px rgba(33, 33, 33, 0.2);
}
