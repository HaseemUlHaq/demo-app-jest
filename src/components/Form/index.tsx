import React, { useState } from 'react';
import './form.css';
import FormHeader, { defaultHeaderProps } from './FormHeader';

export interface FormProps {
  name: string;
  age: number;
}

export const sendToApi = (body: FormProps) => {
  return body;
};

const Form: React.FC = () => {
  const [form, setForm] = useState<FormProps>(defaultHeaderProps.form);

  const handleChange = (e: { target: { name: string; value: string | number } }) => {
    const { name, value } = e.target;
    setForm({ ...form, [name]: value });
  };

  const handleSubmit = (e: any) => {
    e.preventDefault();
    sendToApi(form);
  };

  return (
    <form onSubmit={handleSubmit}>
      <FormHeader form={form} />
      <p>Enter your name:</p>
      <input type="text" name="name" onChange={handleChange} />
      <p>Enter your age:</p>
      <input type="number" name="age" onChange={handleChange} />
      <button title="Submit" type="submit">
        Submit
      </button>
    </form>
  );
};

export default Form;
