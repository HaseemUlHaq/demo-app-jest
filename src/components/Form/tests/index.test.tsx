import React from 'react';
import { mount, shallow } from 'enzyme';
import Form, { sendToApi } from '..';
import toJson from 'enzyme-to-json';

describe('components/Form', () => {
  it('shoudld render without crash', () => {
    shallow(<Form />);
  });

  it('should render with empty fields', () => {
    const wrapper = mount(<Form />);
    const form = wrapper.find('input');

    //name field
    expect(form.at(0).prop('value')).toEqual('');
    //age field
    expect(form.at(1).prop('value')).toEqual(0);
    // header should be initialised by default state
    expect(wrapper.find('FormHeader h1').text()).toEqual('Hello  0');
  });

  it('should create correct snapshot for form', () => {
    const wrapper = mount(<Form />);
    expect(toJson(wrapper.find('form'))).toMatchSnapshot();
  });

  it('should return true when values are passed', () => {
    const body = { name: 'haseem', age: 23 };
    expect(sendToApi(body)).toEqual(body);
  });

  it('should pass correct values to API', () => {
    const wrapper = mount(<Form />);
    wrapper.find('input[name="name"]').simulate('change', {
      target: { name: 'name', value: 'Onyx' },
    });
    wrapper.find('input[name="age"]').simulate('change', {
      target: { name: 'age', value: 10 },
    });
    wrapper.find('form').simulate('submit');
    // expect(form.at(1).prop('name')).toEqual('age');

    expect(wrapper.find('FormHeader h1').text()).toEqual('Hello Onyx 10');
  });

  // send correct data to API
});
