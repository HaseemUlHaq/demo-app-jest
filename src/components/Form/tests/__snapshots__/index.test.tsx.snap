// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form should create correct snapshot for form 1`] = `
<form
  onSubmit={[Function]}
>
  <FormHeader
    form={
      Object {
        "age": 0,
        "name": "",
      }
    }
  >
    <h1>
      Hello 
       
      0
    </h1>
  </FormHeader>
  <p>
    Enter your name:
  </p>
  <input
    name="name"
    onChange={[Function]}
    type="text"
  />
  <p>
    Enter your age:
  </p>
  <input
    name="age"
    onChange={[Function]}
    type="number"
  />
  <button
    title="Submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;
